mlos_bench/mlos_bench/tunables/covariant_group.py

```diff
@@ -179,15 +180,15 @@ def get_names(self) -> Iterable[str]:
         """
         return self._tunables.keys()
 
-    def get_tunable_values_dict(self) -> Dict[str, TunableValue]:
+    def get_tunable_values_dict(self) -> HashableTunableValuesDict:
         """
         Get current values of all tunables in the group as a dict.
 
         Returns
         -------
-        tunables : Dict[str, TunableValue]
+        tunables : HashableTunableValuesDict
         """
-        return {name: tunable.value for (name, tunable) in self._tunables.items()}
+        return HashableTunableValuesDict({name: tunable.value for (name, tunable) in self._tunables.items()})
```

Comment:
I would rather not expose the data structure that exists exclusively for the convenience of the grid search optimizer here in the Tunables or anywhere on the scheduler side. I am not even sure we need that `HashableTunableValuesDict` structure at all - how about we just keep track of the tuples of tunable values in the optimizer? I'll comment on that in detail at `GridSearchOptimizer`
Alternatively, just make `TunableGroups` hashable

Reply:
Yeah, I was waffling on this too.

I think part of the issue is that we are somewhat inconsistent about dicts vs. TunableGroups vs. DataFrames.

I think both of the options you proposed are good though, so I'll play with it.

Reply:
An issue with making TunableGroups, CovariantGroups, Tunable hashable is that they are all mutable so one could do something like this:

```python
configs = set()
tunable_groups = TunableGroup(...)
configs.add(tunable_groups)
tunable_groups.assign(new_values)
assert len(configs) == 1
assert tunable_groups in configs  # would fail, since the hash now no longer matches
```

I think from a data structure perspective this is wrong.

The way I'd written `HashableTunableValuesDict` it was immutable, so this couldn't happen.

I can rework things internally to just use `tuples` of values, but it does make for some inefficiencies (e.g., having to reconstruct `dicts` from those repeatedly), but that might be less important.

Anyways, just leaving a note about this for now.

Reply:
FWIW, using `tuples` is also sensitive to the order they arrive into the `dict` in, which is different between `ConfigSpace` and `TunableGroups`